Store PersistentCollection changes in entity change set as array

franmomu opened the pull request:
The property `$entityChangeSets` of UnitOfWork stores entity changes as an array of two items for each property changed. 

https://github.com/doctrine/orm/blob/a3a8caae516d8b88c025692159ce44f7ee7f1844/lib/Doctrine/ORM/UnitOfWork.php#L143-L149

But for `Collection` it has another behaviour as pointed out in https://github.com/doctrine/orm/discussions/8955 and changed the phpdoc in https://github.com/doctrine/orm/pull/8959.

That was added in https://github.com/doctrine/orm/commit/65bbdc30de891340b685c483046b535be4282718.

This PR modifies that case storing an `array` instead of the `PersistentCollection`, this will allow `PreUpdateEventArgs` to extend from `PreUpdateEventArgs` of `doctrine/persistence`.

I'm not sure if this is the right fix, but apparently the test added in the original commit passes with this change.

greg0ire commented on UPGRADE.md:
```diff
@@ -1,5 +1,11 @@
 # Upgrade to 3.0
 
+## BC BREAK: Changed entity change set when a `Collection` is modified
```
```suggestion
## BC BREAK: Changed entity changeset when a `Collection` is modified
```

greg0ire commented on UPGRADE.md:
```diff
@@ -1,5 +1,11 @@
+
+When working with an entity with a `Collection` property, when that property
+is changed, the changeset stored in `UnitOfWork` includes the old and the new
+value of this property as array of two items.
```
```suggestion
value of this property as an array of two items.
```

SenseException commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -683,7 +683,7 @@ public function computeChangeSet(ClassMetadata $class, object $entity): void
 
                     $this->collectionDeletions[$coid] = $orgValue;
-                    $changeSet[$propName]             = $orgValue; // Signal changeset, to-many assocs will be ignored.
+                    $changeSet[$propName]             = [$orgValue, $actualData];
```
I wonder if this change will leave dead code somewhere that handled the non-array value of `$changeSet[$propName] = $orgValue;`, but I haven't found any.

franmomu committed:
Store PersistentCollection changes in entity change set as array

The property $entityChangeSets of UnitOfWork stores entity changes
as an array of two items for each property changed. Before this
change, when an entity with a property holding a Collection was
changed, $entityChangeSets stored only the old Collection.

derrabus commented:
Can we prepare this change on 2.x somehow, e.g. by allowing apps to opt-in to it?

franmomu committed:
PreUpdateEventArgs inherits from persistence PreUpdateEventArgs event

franmomu commented:
I forgot about this PR.

> Can we prepare this change on 2.x somehow, e.g. by allowing apps to opt-in to it?

I cannot think a way to do that 😞
